cmake/Re2cBuildType.cmake

```diff
@@ -1,14 +1,76 @@
-set(_allowed_build_types Debug Release RelWithDebInfo MinSizeRel)
-
-# 1. The default value for CMAKE_BUILD_TYPE is an empty string.
-# 2. User can set CMAKE_BUILD_TYPE to any value at the cmake command line.
-#
-# Therefore, we check both cases, and make sure that we are dealing with
-# a known build type (if any).
-#
-# For more see discussion at https://github.com/skvadrik/re2c/pull/316
-if(CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE IN_LIST _allowed_build_types)
-    message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
+# CMAKE_BUILD_TYPE is not used by multi-configuration generators.
+# Thus, we need to check this first.
+get_property(isMultiConfig GLOBAL
+    PROPERTY GENERATOR_IS_MULTI_CONFIG)
+
+if(isMultiConfig) # Xcode, Visual Studio, ...
+    if (NOT WIN32)
+        if(NOT "Asan" IN_LIST CMAKE_CONFIGURATION_TYPES)
+            list(APPEND CMAKE_CONFIGURATION_TYPES Asan)
+        endif()
+        if(NOT "Ubsan" IN_LIST CMAKE_CONFIGURATION_TYPES)
+            list(APPEND CMAKE_CONFIGURATION_TYPES Asan)
+        endif()
+    endif()
```

Comment:
Can you change the structure of the code as follows, to simplify the handling of multi-config builds:
```cmake
if(NOT isMultiConfig) # Makefiles, Ninja, ...
    ...
else if(NOT MULTICONFIG_DONE) # Xcode, Visual Studio, ...
    set(MULTICONFIG_DONE TRUE)
    if (NOT WIN32)
        list(APPEND CMAKE_CONFIGURATION_TYPES Asan Ubsan)
    endif()
endif()
```
Also note that line 12 has `Asan` instead of `Ubsan`.

Is multi-configuration build automatically set depending on the platform, or is it something the user can choose as well?